### PR TITLE
c10d/logging: add C10D_LOCK_GUARD

### PR DESCRIPTION
**Pull request by d4l3k**
This adds logs if we can't acquire locks in NCCLUtils and ProcessGroupNCCL for 30s.

This is motivated by some deadlocks were seeing and it's unclear if it's in NCCL or on the PyTorch side of things.

This required replacing most `std::mutex` with `std::timed_mutex` and `std::condition_variable_any` as appropriate.

Test plan:

existing CI for regressions

will add unit tests on `C10D_LOCK_GUARD`

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @c-p-i-o @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10 @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @snadampal

### CODE REVIEW

File: torch/csrc/distributed/c10d/logging.cpp [hunk 1]
```diff
@@ -34,4 +34,20 @@ bool isLogLevelEnabled(LogLevel level) noexcept {
   return false;
 }
 
+void lockWithLogging(
+    std::unique_lock<std::timed_mutex>& lock,
+    std::chrono::milliseconds log_interval,
+    c10::string_view desc,
+    c10::string_view file,
+    int line) {
+  while (!lock.try_lock_for(log_interval)) {
```
Comment (fduwjj) [inline torch/csrc/distributed/c10d/logging.cpp:43]: I am wondering if we can just do `try_lock_for` just once and then try to capture the lock like what we do in the past?

Comment (d4l3k) [reply torch/csrc/distributed/c10d/logging.cpp:43]: yeah, that's fair -- just depends if it's too much log spam

File: torch/csrc/distributed/c10d/Work.cpp [hunk 2]
```diff
@@ -45,17 +46,17 @@ OpType Work::retrieveOpType() const {
...
 
 bool Work::isCompleted() {
-  std::lock_guard<std::mutex> lock(mutex_);
+  C10D_LOCK_GUARD(lock, mutex_);
   return completed_;
 }
 
 bool Work::isSuccess() const {
-  std::lock_guard<std::mutex> lock(mutex_);
+  C10D_LOCK_GUARD(lock, mutex_);
```
Comment (c-p-i-o) [inline torch/csrc/distributed/c10d/Work.cpp:54]: wonder if it makes sense to change the name `lock` to instead be more descriptive?
We can do that later too.

**Review by c-p-i-o — APPROVED** [review]

File: torch/csrc/distributed/c10d/Work.cpp (hunk 2 above)
Comment (d4l3k) [reply torch/csrc/distributed/c10d/Work.cpp:54]: Right now we don't print the lock name -- just the mutex

This variable name only is used when condition variables are used as well

**Review by fduwjj — APPROVED** [review]
Test looks great!

**Comment by xwang233** [comment]
Thanks @ZainRizvi  for reverting the PR. Our build also failed with the header issue. cc @malfet @atalman 

```
#45 84.81 In file included from /usr/local/lib/python3.10/dist-packages/torch/include/torch/csrc/distributed/c10d/NCCLUtils.hpp:17,
#45 84.81                  from /usr/local/lib/python3.10/dist-packages/torch/include/torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:22,
#45 84.81                  from xxx:
#45 84.81 /usr/local/lib/python3.10/dist-packages/torch/include/torch/csrc/distributed/c10d/logging.h:14:10: fatal error: fmt/format.h: No such file or directory
#45 84.81    14 | #include <fmt/format.h>
#45 84.81       |          ^~~~~~~~~~~~~~
#45 84.81 compilation terminated.
```

**Comment by d4l3k** [comment]
@xwang233 you likely need to add fmt as a dependency to your custom build -- we depend on fmt elsewhere in torch but not in torch_cuda/rocm until now.

Will reland this once CI passes

**Comment by d4l3k** [comment]
Ah -- we're leaking fmt into the public headers, will fix

**Comment by kit1980** [comment]
This was reverted by D61985186

**Pull request closed without merging**